### PR TITLE
Bond integration tests & migration

### PR DESCRIPTION
**Pull request by bucurdavid**

### CODE REVIEW

**Comment by newbreedofgeek** [comment]
tnx @bucurdavid 

I can see that we are now using the proxy implementation.

@damienen can you code review pls.

After which i'll upgrade the devnet minter so we can test the new proxy work.

File: src/lib.rs [hunk 1]
```diff
@@ -166,9 +169,13 @@ pub trait DataNftMint:
 
         let treasury_address = self.treasury_address().get();
 
+        let bond_amount = self.get_bond_amount_for_lock_period(lock_period_sec);
+
         if price >= BigUint::zero() {
```
Comment (damienen) [inline src/lib.rs:174]: What's the point of this if? Can't you just do:

require!(
                payment.amount >= &price + &bond_amount,
                ERR_NOT_ENOUGH_FUNDS
            );

File: src/lib.rs [hunk 2]
```diff
@@ -166,9 +169,13 @@ pub trait DataNftMint:
...
-            // @TODO ideally here we require that the payment is price + bondAmount (implement proxy view for bondAmount from the bonding contract)
-            // ... for now, the TX will fail if the payment.amount does not cover the price (anti_spam_tax) + bondAmount
+            require!(
+                payment.amount >= &price + &bond_amount,
```
Comment (damienen) [inline src/lib.rs:176]: Why is it >= here, but = in the lower else?

File: src/lib.rs [hunk 3]
```diff
@@ -178,6 +185,8 @@ pub trait DataNftMint:
             );
 
             payment.amount -= &price;
+        } else {
+            require!(payment.amount == bond_amount, ERR_NOT_ENOUGH_FUNDS);
```
Comment (damienen) [inline src/lib.rs:189]: here is == (and upper is >=)

File: tests/endpoints/collection_management.rs [hunk 4]
```diff
@@ -0,0 +1,4 @@
+#[test]
+fn collection_management_test() {
```
Comment (damienen) [inline tests/endpoints/collection_management.rs:2]: Shouldn't these get implemented as well?

File: tests/endpoints/mint.rs [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+        "random-url-encoded-here",
+        "https://test.com/test",
+        1000u64,
+        5u64,
+        &"Test title".repeat(1),
+        &"Test description".repeat(1),
+        10u64,
+        ITHEUM_TOKEN_IDENTIFIER,
+        0u64,
+        100u64 + 100u64,
```
Comment (damienen) [inline tests/endpoints/mint.rs:451]: Shouldn't you have tests that check this?
require!(
payment.amount >= &price + &bond_amount,
ERR_NOT_ENOUGH_FUNDS
);

You are only checking if all is well when sending equal or lower funds. What if anti spam tax is 0 and you send equal AND more funds (separate cases). What if anti spam tax is not 0 and you send equal AND more funds (separate cases)?

**Review by damienen — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (bucurdavid) [reply src/lib.rs:176]: It should be ==, thanks for the catch

File: src/lib.rs (hunk 1 above)
Comment (bucurdavid) [reply src/lib.rs:174]: If payment is 0, it shouldn't send funds to treasury address, if the check is not here will be somewhere else in the code ..

File: tests/endpoints/collection_management.rs (hunk 4 above)
Comment (bucurdavid) [reply tests/endpoints/collection_management.rs:2]: These functionalities are not implemented in the testing framework. As far as I know, they were also not available in the older framework.

Comment (damienen) [reply tests/endpoints/collection_management.rs:2]: Makes sense

File: src/lib.rs (hunk 1 above)
Comment (damienen) [reply src/lib.rs:174]: Alright. Makes sense. Then I guess the check for equality can be gotten out of the if.

Also, as a cool thing (not mandatory to do) you can user direct_non_zero to catch the case in which the transfer is zero

File: tests/endpoints/mint.rs (hunk 5 above)
Comment (bucurdavid) [reply tests/endpoints/mint.rs:451]: It should be 
```suggestion
require!(
payment.amount == &price + &bond_amount,
ERR_NOT_ENOUGH_FUNDS
);
```

Comment (bucurdavid) [reply tests/endpoints/mint.rs:451]: The check you are asking is in the same file under this test `mint_test_without_anti_spam_tax_test`

Comment (damienen) [reply tests/endpoints/mint.rs:451]: Cool then, my bad

**Review by damienen — APPROVED** [review]

**Review by damienen — APPROVED** [review]

**Review by newbreedofgeek — APPROVED** [review]

**Pull request merged**